### DISCUSSION TITLE
Recommended pattern for required cookies

### DISCUSSION THREAD

**Discussion by saevarb** [fastapi/fastapi#11308] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
CookieDep = Annotated[str | None, Cookie()]
@router.get("/")
async def get_session(session_id: CookieDep = None) -> ChatSession | None:
    return SessionModel.get_session(session_id)
```

### Description

Hi.

I'm using cookies for basic sessions and I'm struggling to figure out how to have required cookies that don't automatically(and incorrectly) get added as explicit parameters in the swagger UI.

A cookie is an implicit parameter. It's either present or it isn't.

Using code like the following

```
CookieDep = Annotated[str | None, Cookie()]
@router.get("/")
async def get_session(session_id: CookieDep = None) -> ChatSession | None:
    return SessionModel.get_session(session_id)
```

lets me not pass the cookie when using the `/docs` endpoint, but it also tells FastAPI it's not required, and requires additional middleware or something like that for me to verify that it's present.

If I instead change it to

```
CookieDep = Annotated[str | None, Cookie()]
@router.get("/")
async def get_session(session_id: CookieDep) -> ChatSession | None:
    return SessionModel.get_session(session_id)
```

Then it is correctly required by FastAPI, but then I need to explicitly pass it as an argurment in the `/docs` endpoint, which is not correct. It would be OK if passing the cookie manually like that was optional(so I could set any session ID by putting it in the input for testing purposes), but trying to execute the endpoint without putting anything in the (now required) cookie parameter in `/docs` leads to a runtime crash in the browser.

### Operating System

Linux

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

3.11.7

**Comment by YuriiMotov** [comment, marked as answer]
If you don't care about the consistency of your OpenAPI schema, you can modify it to make `session_id` not required. So, documentation systems will think that it's not required and allow you leave this field empty, but validation will still work on the backend side.
```
from typing import Annotated

from fastapi import Cookie, FastAPI
from fastapi.openapi.utils import get_openapi

app = FastAPI(swagger_ui_parameters={"validatorUrl": None})

CookieDep = Annotated[str, Cookie()]
@app.get("/")
async def get_session(session_id: CookieDep) -> None:
    pass

def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        summary=app.summary,
        description=app.description,
        routes=app.routes,
    )
    for path_methods in openapi_schema["paths"].values():
        for operation in path_methods.values():
            for param in operation["parameters"]:
                if param["name"] == "session_id":
                    param["required"] = False

    app.openapi_schema = openapi_schema
    return app.openapi_schema

app.openapi = custom_openapi
```

I also thought that it's possible to disable the Browser-side request validation in Swagger, but it seems like there is no such an option..

You can also think about creating two different `OpenAPI.json` endpoints: one for Swagger\Readoc, and another one for developers. So you can make this parameters not required only in first `OpenAPI.json` and leave them required in the second.
